client/src/components/exercise/lab8/pages/DataRepair.js

```diff
@@ -0,0 +1,393 @@
+          ? { ...message, ai_polarity: newValue }
+          : message
+      )
+    );
+  };
+  /*
+    make sure that users cannot click "previous" or "continue buttons"
+    while they are playing the exercise
+    */
+  useEffect(() => {
+    actions.updateState(EXERCISE_PLAYING);
+  }, [actions]);
+
+  // the only acceptable values that a user can enter
+  // for their repairs
+  const repairAllowList = [0, 1, 2];
+
+  /*
+    set the message to be displayed in the popup
+  */
+  const popUpHandler = (message) => {
+    setPopUpMessage(message);
+  };
+
+  /*
+    confirm that the user entered a value for every message
+    confirm that the user entered either 0, 1, or 2 for every message
+    i.e. they cannot enter a string, boolean, etc.
+    use the allow list to compare user entered value to acceptable values
+    */
+  const validateRepair = () => {
+    // track if the user made an error in their repairs
+    let error = false;
+    const localMessageError = [...messageError];
+
+    messages.forEach((message, index) => {
+      if (!(message.ai_polarity in repairAllowList)) {
+        // we need to display an error message
+        error = true;
+        localMessageError.splice(index, 1, true);
+      } else {
+        // clear the error message
+        localMessageError.splice(index, 1, false);
+      }
+    });
+    if (!error) {
+      // eventually need to send repair data to the backend
+      setRepairOpen(false);
+      setUserError(false);
+      popUpHandler("The repairs have been made.");
+      setRepairState(true);
+      setIsCorrect(validateCorrectAI());
+    } else {
+      setUserError(true);
+      popUpHandler(
+        "There are errors in your repair. Please correct the errors."
+      );
+    }
+    setMessageError(localMessageError);
+  };
+
+  /**
+   * fetchDataRepair(): is a function responsible for handling retrieval of data to
+   * populate data repair sections.
+   * @returns returns object containing user lab information
+   */
+  const fetchDataRepair = async () => {
+    try {
+      const { userid } = user;
+      return await ExerciseService.getUserRepair(userid);
+    } catch (error) {
+      console.error(error);
+    }
+  };
+  /**
+   * postDataRepair(): is a function responsible for handling posting of data to
+   * populate data repair sections.
+   * @returns returns object containing user lab information
+   */
+  const postExerciseChange = async (data) => {
+    try {
+      return await ExerciseService.submitRepair(data);
+    } catch (error) {
+      console.error(error);
+    }
+  };
+
+  /**
+   * handleRepair(): is a helper function that is responsible for the
+   * opening and closing of the repair window in this view.
+   */
+  const handleRepair = async () => {
+    !repairOpen ? setRepairOpen(true) : "";
+    const dataRepair = await fetchDataRepair();
+    if (dataRepair?.userid) {
+      const { repair, numRepair, isComplete } = dataRepair;
+      isComplete
```

Comment:
i don't believe we need the spread operator. it should be used when we are combining multiple arrays/lists, but since it's only one list, we should just set the message state to the data set we want. 

`setMessages(isComplete ? CHAT_MESSAGES.messages : repair.messages)`

Reply:
Good point!